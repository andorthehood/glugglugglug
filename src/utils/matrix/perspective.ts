export function perspective(
	fieldOfViewInRadians: number,
	aspect: number,
	near: number,
	far: number,
) {
	const f = Math.tan(Math.PI * 0.5 - 0.5 * fieldOfViewInRadians);
	const rangeInv = 1.0 / (near - far);

	return [
		f / aspect,
		0,
		0,
		0,
		0,
		f,
		0,
		0,
		0,
		0,
		(near + far) * rangeInv,
		-1,
		0,
		0,
		near * far * rangeInv * 2,
		0,
	];
}
