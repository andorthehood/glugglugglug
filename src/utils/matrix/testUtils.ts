export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
