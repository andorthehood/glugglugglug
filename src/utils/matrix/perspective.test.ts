import { describe, expect, it } from "vitest";
import { perspective } from "./perspective";
import { formatMatrix } from "./testUtils";

describe("perspective", () => {
	it("builds a perspective projection matrix", () => {
		const fieldOfViewInRadians = Math.PI / 3;
		const aspect = 16 / 9;
		const near = 1;
		const far = 200;

		const result = perspective(fieldOfViewInRadians, aspect, near, far);

		expect(formatMatrix(result)).toMatchSnapshot();
	});
});
