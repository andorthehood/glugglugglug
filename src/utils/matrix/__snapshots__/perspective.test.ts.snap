// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`perspective > builds a perspective projection matrix 1`] = `
[
  [
    0.974279,
    0,
    0,
    0,
  ],
  [
    0,
    1.732051,
    0,
    0,
  ],
  [
    0,
    0,
    -1.01005,
    -1,
  ],
  [
    0,
    0,
    -2.01005,
    0,
  ],
]
`;
